### PR TITLE
Pass to reduce added flags from RemoveReturn

### PR DESCRIPTION
**Pull request by ChrisDodd**
This adds a pass running before RemoveReturns that rewrites code with a conditional jump (a break/continue/exit/return in a if branch) where there's code after the if statement that will only run if the if condition is false.  It moves that following code into the else (ifFalse) part of the IfStatement

The point of this is avoid needing to introduce a flag variable in the later RemoveReturns or RemoveExits.  It can also help with loops.

### CODE REVIEW

**Comment by jafingerhut** [comment]
@fruffy Does something like gauntlet run automatically on all PRs?  Or is there a way to manually trigger it on a particular PR by adding a tag?  This PR seems likes a good one to have that extra checking, if it is easy to enable.

**Comment by fruffy** [comment]
> @fruffy Does something like gauntlet run automatically on all PRs? Or is there a way to manually trigger it on a particular PR by adding a tag? This PR seems likes a good one to have that extra checking, if it is easy to enable.

Yes, I added the `run-validation` tag which will run the tool. It will also run nightly after it has been merged.

**Review by jafingerhut — APPROVED** [review]
Caveats: I did not review the C++ code changes.  I _also_ did not carefully review most of the expected output P4C changes.  They are simply too long to do so in a reasonable amount of time, hence my question of Fabian about adding the run-validation tag to get better automated coverage of the correctness of those changes.

Thus this is a very "weak" LGTM approval, based upon me manually checking a few of the very short expected output files, and finding no problems.

If someone is willing to review parts of this PR that I have not, I would consider my approval to be not enough on its own to merge this.

**Comment by ChrisDodd** [comment]
Turns out there's a bug here (incorrectly applying changes) due to ignoring the effects of switch statements (and loops), so I'll push a revision soon.

File: frontends/p4/removeReturns.h
Comment (vlstill) [inline frontends/p4/removeReturns.h:45]: By `branch` you presumably mean `exit`, `return`, `continue`, right? Please add that to the comment as when `branch` is said, I imagine `if` or `switch`.

File: frontends/p4/removeReturns.h
Comment (vlstill) [inline frontends/p4/removeReturns.h:71]:
```suggestion
     * not across calls), and we only have one bit of data (hasBranched flag). */

File: frontends/p4/removeReturns.cpp
Comment (vlstill) [inline frontends/p4/removeReturns.cpp:91]: There can't be `exit` in a loop? I.e. 

```p4
for (bit<4> x in ...) {
   if (x > y) {
      exit;
   }
}
```

or even degenerated loop:

```p4
for (...) {
    exit;
}
```

I guess the point is that if the loop can not run even once than it will never fire the exit, right?

**Review by vlstill — COMMENTED** [review]
I like this. I've noticed the `RemoveRetun` pass introduces way too much flags, but did not get to figuring out why and what we can do about this :-). I just have some questions.

File: frontends/p4/removeReturns.cpp
Comment (ChrisDodd) [reply frontends/p4/removeReturns.cpp:91]: The loop body might not run at all, so any exit in the loop is by definition not unconditional.

File: frontends/p4/removeReturns.h
Comment (ChrisDodd) [reply frontends/p4/removeReturns.h]: Yes, this is confusing.  I rewrote the comment to use the term "jump statement" (as that is what the C spec uses for these) instead of "branch"

**Review by vlstill — APPROVED** [review]
Maybe just also rename `hasBranched` to `hasJumped` or something like that.

File: frontends/p4/removeReturns.h
Comment (vlstill) [reply frontends/p4/removeReturns.h]: Thansk, that is a better name :-).

File: frontends/p4/removeReturns.h
Comment (vlstill) [inline frontends/p4/removeReturns.h:78]: Why is the comment after the member? Might be best to put it before as a doxygen comment (`/// ...`).

**Pull request merged**
